source/github-helpers/is-conversation-locked.ts

```diff
@@ -0,0 +1,53 @@
+import elementReady from 'element-ready';
+
+import {isInitialLoad} from '../helpers/feature-helpers.js';
+import {hasToken} from '../options-storage.js';
+import api from '../github-helpers/api.js';
+import GetIssueLockStatus from './locked-issue.gql';
+import {getConversationNumber} from '../github-helpers/index.js';
+
+async function isConversationLockedViaApi(): Promise<boolean | undefined> {
+	if (!hasToken()) {
+		return undefined;
+	}
+
+	const {repository} = await api.v4uncached(GetIssueLockStatus, {
+		variables: {
+			number: getConversationNumber()!,
+		},
+	});
+
+	return repository.issueOrPullRequest.locked;
+}
+
+async function isConversationLockedViaDom(): Promise<boolean | undefined> {
+	// Only use signals that clearly indicate the lock state
+	// The form only appears to moderators
+	const lockToggle = await elementReady('.discussion-sidebar-item svg.octicon-key + strong');
+	return lockToggle ? lockToggle.textContent === 'Unlock conversation' : undefined;
+}
+
+async function isConversationLockedViaReactData(): Promise<boolean | undefined> {
+	if (!isInitialLoad()) {
+		return;
+	}
+
+	const data = await elementReady('[data-target="react-app.embeddedData"]');
+	return data ? JSON.parse(data.textContent).payload?.preloadedQueries[0].result.data.repository?.issue?.locked : undefined;
+}
+
+export default async function isConversationLocked(): Promise<boolean | undefined> {
+	// Like Promise.race, but it only resolves if the result is not undefined
+	return new Promise(resolve => {
+		// TODO: Add AbortSignal after https://github.com/sindresorhus/element-ready/issues/45
+		const resolveIfDefined = async (check: () => Promise<boolean | undefined>): Promise<void> => {
+			const result = await check();
+			if (result !== undefined) {
+				resolve(result);
+			}
+		};
+		resolveIfDefined(isConversationLockedViaReactData);
+		resolveIfDefined(isConversationLockedViaDom);
+		resolveIfDefined(isConversationLockedViaApi);
```

Comment:
I don't really like that we have 3 ways to do one thing, but alas the API call is always 300/700ms slower than the DOM :(